### PR TITLE
[FEATURE] Identify: more selection modes (simple / polygon / freehand / radius)

### PR DESCRIPTION
**Pull request by wonder-sk**
This replaces PR #6639

Feature request: https://issues.qgis.org/issues/18224

Extended identify map tool where user can choose among multiple options to identify a feature. Beside currently supported single click, new options are available similar to ones selection map tool has (e.g identify within a radius, polygon or rectangular area).

Identify menu

Thanks to **East Hampshire District Council** and **Leicester County Council** for funding this feature!

### CODE REVIEW

**Comment by nyalldawson** [comment]
Thanks for these revisions! I'll review again when I get some time. 

Just wondering though from the screenshot - shouldn't that be "identify features by..." instead of "select features by..."?

**Comment by wonder-sk** [comment]
Good point with naming - fixed now and screenshot updated.

**Comment by wonder-sk** [comment]
Here's my small experiment with the polygon mode icon: no yellow rect / original / with blue circle

image

My preference would be probably towards the left icon, but I am no designer :-)

**Comment by wonder-sk** [comment]
@nirvn @DelazJ I would appreciate your feedback on the comment above :-)

**Comment by nirvn** [comment]
@wonder-sk , oups, sorry. My preference would be the right icon, but with the blue circle _above_ the dash outline.

**Comment by DelazJ** [comment]
Sorry @wonder-sk  Agreed with @nirvn The blue is more meaningful and wins imho

**Comment by wonder-sk** [comment]
Thanks for your inputs - my original thinking was that because of the fact these icons are within the "identify results" dock widget, people know that the icon is related to identification, so the blue circle is not needed... Anyway I will update the icons with the blue circle on top as @nirvn suggested.

**Comment by m-kuhn** [comment]
I also have an aversion to the blue info circle, but I think it's mainly because it doesn't blend very well with the rest of the flat icons.

File: src/app/qgsidentifyresultsdialog.h [hunk 1]
```diff
@@ -167,6 +168,8 @@ class APP_EXPORT QgsIdentifyResultsDialog: public QDialog, private Ui::QgsIdenti
      */
     QgsExpressionContextScope expressionContextScope() const;
 
+    QgsMapToolSelectionHandler::SelectionMode selectionMode();
```
Comment (m-kuhn) [inline src/app/qgsidentifyresultsdialog.h]: const?

File: src/gui/qgsmaptoolselectionhandler.cpp [hunk 2]
```diff
@@ -0,0 +1,460 @@
...
+}
+
+QgsMapToolSelectionHandler::~QgsMapToolSelectionHandler()
+{
+  cancel();
+}
+
+void QgsMapToolSelectionHandler::setInterface( QgisInterface *iface )
+{
+  mQgisInterface = iface;
```
Comment (m-kuhn) [inline src/gui/qgsmaptoolselectionhandler.cpp]: What's your opinion on QgisInterface usage in gui classes?
I would prefer a solution with either this class moved to app or with iface interaction delegated to calling code.

File: src/gui/qgsmaptoolselectionhandler.cpp [hunk 3]
```diff
@@ -0,0 +1,460 @@
...
+/// @endcond
+
+
+
+QgsMapToolSelectionHandler::QgsMapToolSelectionHandler( QgsMapCanvas *canvas, QgsMapToolSelectionHandler::SelectionMode selectionMode )
+  : QObject()
+  , mCanvas( canvas )
+  , mSelectionMode( selectionMode )
+  , mSnapIndicator( qgis::make_unique< QgsSnapIndicator >( canvas ) )
+  , mFillColor( QColor( 254, 178, 76, 63 ) )
```
Comment (m-kuhn) [inline src/gui/qgsmaptoolselectionhandler.cpp]: Can this be moved to the header?

File: src/gui/qgsmaptoolselectionhandler.cpp (hunk 2 above)
Comment (wonder-sk) [reply src/gui/qgsmaptoolselectionhandler.cpp]: Actually only now I have realized that after refactoring this class is not needed in gui anymore - let me move it app then...

**Comment by nirvn** [comment]
@wonder-sk , great. Bonus points if you create a clear identify result icon using blue circle on top left and the red glyph on bottom right. :)

Thanks for updating the icons.

File: src/app/qgsidentifyresultsdialog.h [hunk 4]
```diff
@@ -251,6 +256,7 @@ class APP_EXPORT QgsIdentifyResultsDialog: public QDialog, private Ui::QgsIdenti
     QList<QgsFeature> mFeatures;
     QMap< QString, QMap< QString, QVariant > > mWidgetCaches;
     QgsExpressionContextScope mExpressionContextScope;
+    QToolButton *mSelectModeButton;
```
Comment (nyalldawson) [inline src/app/qgsidentifyresultsdialog.h]: = nullptr

File: src/app/qgsmaptoolselectionhandler.h [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+    //! the rubberband for selection visualization
+    std::unique_ptr<QgsRubberBand> mSelectionRubberBand;
+
+    //! Stores the most recent geometry (in map coordinates)
+    QgsGeometry mSelectionGeometry;
+
+    //! Whether a geometry is being picked right now
+    bool mSelectionActive = false;
+
+    SelectionMode mSelectionMode;
```
Comment (nyalldawson) [inline src/app/qgsmaptoolselectionhandler.h]: initialize

File: src/gui/qgsmaptoolidentify.cpp [hunk 6]
```diff
@@ -221,21 +258,31 @@ bool QgsMapToolIdentify::identifyVectorLayer( QList<IdentifyResult> *results, Qg
...
+          selectionGeom.transform( ct );
+      }
+    }
 
     QgsFeatureIterator fit = layer->getFeatures( QgsFeatureRequest().setFilterRect( r ).setFlags( QgsFeatureRequest::ExactIntersect ) );
     QgsFeature f;
     while ( fit.nextFeature( f ) )
-      featureList << QgsFeature( f );
+    {
+      if ( isPointOrRectangle || selectionGeom.intersects( f.geometry() ) )
```
Comment (nyalldawson) [inline src/gui/qgsmaptoolidentify.cpp]: I think you should use a prepared geos geometry here for the intersection test - it's going to make this much faster.

**Review by nyalldawson — APPROVED** [review]
Looks fantastic to me -- very clean approach!

**Pull request merged**